dandi/pynwb_utils.py

```diff
@@ -262,9 +263,84 @@ def _get_pynwb_metadata(path: str | Path | Readable) -> dict[str, Any]:
         # get external_file data:
         out["external_file_objects"] = _get_image_series(nwb)
 
+        # Calculate session duration for metadata
+        session_duration = _get_session_duration(nwb)
+        if session_duration is not None and out.get("session_start_time") is not None:
+            # Convert to absolute datetime by adding duration to session_start_time
+            start_time = out["session_start_time"]
+            out["session_end_time"] = start_time + timedelta(seconds=session_duration)
+
     return out
 
 
+def _get_session_duration(nwb: pynwb.NWBFile) -> float | None:
+    """Calculate the duration of a recording session from NWB file contents.
+
+    This function finds the minimum and maximum timestamps across all TimeSeries
+    and DynamicTable objects with time information, then returns the duration as
+    max - min.
+
+    Parameters
+    ----------
+    nwb: pynwb.NWBFile
+        An open NWB file object
+
+    Returns
+    -------
+    float or None
+        The session duration in seconds (max_time - min_time),
+        or None if no time information could be extracted
+    """
+    start_times: list[float] = []
+    end_times: list[float] = []
+
+    # Iterate through all objects in the NWB file
+    for obj in nwb.objects.values():
+        try:
+            # Handle TimeSeries objects
+            if isinstance(obj, pynwb.base.TimeSeries):
+                if obj.timestamps is not None and len(obj.timestamps) > 0:
+                    # Use first and last timestamps
+                    start_times.append(float(obj.timestamps[0]))
+                    end_times.append(float(obj.timestamps[-1]))
+                elif (
+                    obj.starting_time is not None
+                    and obj.rate is not None
+                    and obj.data is not None
+                ):
+                    # Calculate start and end time
+                    start_times.append(float(obj.starting_time))
+                    num_samples = len(obj.data)
+                    end_times.append(
+                        float(obj.starting_time + (num_samples / obj.rate))
+                    )
+
+            # Handle DynamicTable objects with start_time and stop_time
+            elif isinstance(obj, hdmf.common.DynamicTable):
+                if "start_time" in obj.colnames:
+                    start_data = obj["start_time"].data[:]
+                    if len(start_data) > 0:
+                        start_times.append(float(min(start_data)))
+                if "stop_time" in obj.colnames:
+                    stop_data = obj["stop_time"].data[:]
+                    if len(stop_data) > 0:
+                        end_times.append(float(max(stop_data)))
```

Comment:
could be shortened even further but at least let's do

```suggestion
                if "start_time" in obj.colnames:
                    if len(start_data := obj["start_time"].data):
                        start_times.append(float(min(start_data)))
                if "stop_time" in obj.colnames:
                    if len(stop_data := obj["stop_time"].data):
                        end_times.append(float(max(stop_data)))
```

unless there was some meaning for explicit slicing with `[:]`, then those should get back I guess